simulation/test_http_connection.cpp

```diff
@@ -438,6 +449,50 @@ TORRENT_TEST(http_connection_socks5_proxy_names)
 	run_suite(ps);
 }
 
+// tests the error scenario of a http server listening on two sockets (ipv4/ipv6) which
+// both accept the incoming connection but never send anything back. we test that
+// both ip addresses get tried in turn and that the connection attempts time out as expected.
+TORRENT_TEST(http_connection_timeout)
+{
+	sim_config network_cfg;
+	sim::simulation sim{network_cfg};
+	// server has two ip addresses (ipv4/ipv6)
+	sim::asio::io_service server_ios(sim, address_v4::from_string("10.0.0.2"));
+	sim::asio::io_service server_ios_ipv6(sim, address_v6::from_string("ff::dead:beef"));
+	// same for client
+	sim::asio::io_service client_ios(sim, {
+		address_v4::from_string("10.0.0.1"),
+		address_v6::from_string("ff::abad:cafe")
+	});
+	lt::resolver resolver(client_ios);
+
+	const unsigned short http_port = 8080;
+	sim::http_server http(server_ios, http_port);
+	sim::http_server http_ipv6(server_ios_ipv6, http_port);
+
+	http.register_stall_handler("/timeout");
+	http_ipv6.register_stall_handler("/timeout");
+
+	char data_buffer[4000];
+	std::generate(data_buffer, data_buffer + sizeof(data_buffer), &std::rand);
+
+	int connect_counter = 0;
+	int handler_counter = 0;
+
+	error_condition timed_out(boost::system::errc::timed_out, boost::system::generic_category());
+
+	auto c = test_request(client_ios, resolver
+		, "http://dual-stack.test-hostname.com:8080/timeout", data_buffer, -1, -1
+		, timed_out, lt::aux::proxy_settings()
+		, &connect_counter, &handler_counter);
+
+	error_code e;
+	sim.run(e);
+	TEST_CHECK(!e);
+	TEST_EQUAL(2, connect_counter); // both endpoints are connected to
```

Comment:
not sure if there's an easy way to test that different endpoints got tried here – maybe we could add some counters to the fake http_servers so one could say
```c++
TEST_EQUAL(1, http.count());
TEST_EQUAL(1, http_ipv6.count());
```